src/tbb/arena.h

```diff
@@ -433,6 +528,9 @@ void arena::advertise_new_work() {
             workers_delta = 1;
         }
 
+#if __TBB_PREVIEW_PARALLEL_PHASE
+        my_thread_leave.reset_if_needed();
```

Comment:
Will the following assert be useful?
```suggestion
        __TBB_ASSERT(mandatory_delta > 0 || workers_delta > 0, "Potentially delaying fast leave?");
        my_thread_leave.reset_if_needed();
```

Reply:
It is not a valid assert. In case of workerless arena and spawn of the task both of those values will be zeroes.

Reply:
Okay... let it be then:
```suggestion
        __TBB_ASSERT((mandatory_delta > 0 || workers_delta > 0) ||
                     (is_arena_workerless() && work_type == work_spawned),
                     "Potentially delaying fast leave?");
        my_thread_leave.reset_if_needed();
```

Reply:
First of all, I do not readily see how this assert is related to the parallel phase.
And if you move the assertion next to the code in the lines 522-529, it seems it just duplicates almost exactly the conditions and assignments there.

Reply:
Yeah, I know that it duplicates. But the stress here is on the words "almost exactly". With the assert I sort of wanted to make sure we do not fall into recently discovered issue of dropping the fast leave setting when threads are recalled from arena. I don't mind not having an assert here, since, as you said, the conditions duplicate it. However, I do think we need to tell reader about the desired behavior here, it feels like the code itself might not be good at telling that.

Another interesting consideration is to reduce cross-NUMA accesses as much as possible. If thread can base its decision whether it needs to read and/or modify global-to-arena variables on the data it has already read, then it makes sense to avoid doing more re-reads of other global-to-arena data. Of course, if we are sure that the set of "almost exactly" is empty here, then the thread already avoids unnecessary touches to the data that increases cross-NUMA traffic.

Reply:
Assertions are good to enforce preconditions for a function, and in some cases of sophisticated internal logic also good to check expected invariants in certain point. Having an assertion that checks the conditions explicitly set by the code right above is a pure duplication. The part that is not an exact duplication could be checked in a proper place, of course; also the `reset_if_needed` could internally assert on its expected preconditions.

For the NUMA part of the comment, I cannot make the connection to the assertion being discussed.